### PR TITLE
Sandy Thomson Chapter 6 Answers

### PR DESCRIPTION
**Pull request by SandyThomson**

### CODE REVIEW

File: src/com/selftest/Stack.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package com.selftest;
+
+public class Stack {
+
+	private char[] rawStack;
+	private int maxArrayIndex;
```
Comment (SimonGross) [inline src/com/selftest/Stack.java:6]: You shouldn't need to store this separately, as you can just call rawStack.length. Although, storing it this way might be easier given the way you're putting things in the array.

File: src/com/selftest/Stack.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+		currentSize = 0;
+		rawStack = new char[initialSize];
+	}
+
+	public void push(char c) {
+		if (currentSize > maxArrayIndex) {
+			throw new RuntimeException("Stack is full, mkay");
+		}
+
+		rawStack[maxArrayIndex - currentSize] = c;
```
Comment (SimonGross) [inline src/com/selftest/Stack.java:20]: Interesting. If I've got this right you're storing from the end of the array backwards. I think storing things from the beginning of the array is simpler, but this works as well

File: src/com/selftest/StringBackwards.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package com.selftest;
+
+public class StringBackwards {
+
+	public static String reverse(String s) {
+
+		if (s.length() > 1) {
+			String stringMinusLastChar = s.substring(0, s.length() - 1);
+			String lastChar = s.substring(s.length() - 1, s.length());
```
Comment (SimonGross) [inline src/com/selftest/StringBackwards.java:9]: Alternatively, as we're just getting a single character you can use `s.charAt( s.length - 1 )`

File: src/com/selftest/Stack.java (hunk 1 above)
Comment (SandyThomson) [inline src/com/selftest/Stack.java:6]: Yeah, I wasn't sure if an array _primitive_ stored the length or not?

File: src/com/selftest/Stack.java (hunk 2 above)
Comment (SandyThomson) [inline src/com/selftest/Stack.java:20]: Yeah, don't know why I did it like this. You are probably right, going the other way would probably be more straightforward. For some reason this felt more intuitive to me, but I can't defend it either.

File: src/com/selftest/StringBackwards.java (hunk 3 above)
Comment (SandyThomson) [inline src/com/selftest/StringBackwards.java:9]: +1 for the protip :smile:

File: src/com/selftest/Stack.java (hunk 1 above)
Comment (utopia-2) [inline src/com/selftest/Stack.java:6]: Arrays of primitives are still objects in Java. They don't provide a whole lot, but you do get a `public final int length` for this purpose.

File: src/com/selftest/Stack.java (hunk 2 above)
Comment (utopia-2) [inline src/com/selftest/Stack.java:20]: You could totally argue this models the stack direction of most CPU architectures better ([SO answer](http://stackoverflow.com/a/664779)).

**Pull request closed without merging**